File: kneron_plus_v2.3.0/kneron_plus/examples/scan_devices/scan_devices.c
/**
 * @file        scan_devices.c
 * @brief       example of scanning Kneron KL series device
 * @version     0.1
 * @date        2021-03-22
 *
 * @copyright   Copyright (c) 2021 Kneron Inc. All rights reserved.
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "kp_core.h"

int main(int argc, char *argv[])
{
    printf("\n");
    printf("scanning kneron devices ...\n");

    kp_devices_list_t *list;
    list = kp_scan_devices();

    printf("number of Kneron devices found: %d\n", list->num_dev);

    if (list->num_dev == 0)
        return 0;

    printf("\n");
    printf("listing devices infomation as follows:\n");

    for (int i = 0; i < list->num_dev; i++)
    {
        kp_device_descriptor_t *dev_descp = &(list->device[i]);

        printf("\n");
        printf("[%d] scan_index: '%d'\n", i, i);
        printf("[%d] port ID: '%d'\n", i, dev_descp->port_id);
        printf("[%d] product_id: '0x%x'", i, dev_descp->product_id);

        if (dev_descp->product_id == KP_DEVICE_KL520)
            printf(" (KL520)");
        else if (dev_descp->product_id == KP_DEVICE_KL720)
            printf(" (KL720)");
        else if (dev_descp->product_id == KP_DEVICE_KL630)
            printf(" (KL630)");
        else if (dev_descp->product_id == KP_DEVICE_KL730)
            printf(" (KL730)");
        else if (dev_descp->product_id == KP_DEVICE_KL830)
            printf(" (KL830)");
        else
            printf(" (Unknown)");
        printf("\n");

        printf("[%d] USB link speed: ", i);
        switch (dev_descp->link_speed)
        {
        case KP_USB_SPEED_LOW:
            printf("'Low-Speed'\n");
            break;
        case KP_USB_SPEED_FULL:
            printf("'Full-Speed'\n");
            break;
        case KP_USB_SPEED_HIGH:
            printf("'High-Speed'\n");
            break;
        case KP_USB_SPEED_SUPER:
            printf("'Super-Speed'\n");
            break;
        default:
            printf("'Unknown'\n");
            break;
        }
        printf("[%d] USB port path: '%s'\n", i, dev_descp->port_path);
        printf("[%d] kn_number: '0x%8X' %s\n", i, dev_descp->kn_number, dev_descp->kn_number == 0x0 ? "(invalid)" : "");
        printf("[%d] Connectable: '%s'\n", i, dev_descp->isConnectable == true ? "True" : "False");
        printf("[%d] Firmware: '%s'\n", i, dev_descp->firmware);
    }

    return 0;
}

File: kneron_plus_v2.2.0/kneron_plus/python/example/ScanDevices.py
# ******************************************************************************
#  Copyright (c) 2021-2022. Kneron Inc. All rights reserved.                   *
# ******************************************************************************

import os
import sys

PWD = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(1, os.path.join(PWD, '..'))

import kp


def get_product_name(product_id: int) -> str:
    if product_id == kp.ProductId.KP_DEVICE_KL520.value:
        return '(KL520)'
    elif product_id == kp.ProductId.KP_DEVICE_KL720.value:
        return '(KL720)'
    elif product_id == 0x200:
        return '(KL720) (Please update firmware by Kneron DFUT)'
    elif product_id == kp.ProductId.KP_DEVICE_KL630.value:
        return '(KL630)'
    else:
        return '(Unknown Device)'


if __name__ == "__main__":
    print('scanning kneron devices ...')

    device_list = kp.core.scan_devices()

    print('number of Kneron devices found: {}'.format(device_list.device_descriptor_number))

    if device_list.device_descriptor_number == 0:
        exit(0)

    print('listing devices infomation as follows:')

    for idx, device_descriptor in enumerate(device_list.device_descriptor_list):
        print()
        print('[{}] USB scan index: \'{}\''.format(idx, idx))
        print('[{}] USB port ID: \'{}\''.format(idx, device_descriptor.usb_port_id))
        print('[{}] Product ID: \'0x{:X} {}\''.format(idx,
                                                      device_descriptor.product_id,
                                                      get_product_name(product_id=device_descriptor.product_id)))

        print('[{}] USB link speed: \'{}\''.format(idx, device_descriptor.link_speed))
        print('[{}] USB port path: \'{}\''.format(idx, device_descriptor.usb_port_path))
        print('[{}] KN number: \'0x{:X}\''.format(idx, device_descriptor.kn_number))
        print('[{}] Connectable: \'{}\''.format(idx, device_descriptor.is_connectable))
        print('[{}] Firmware: \'{}\''.format(idx, device_descriptor.firmware))
